[ResourceMonitor] Add support of monitoring for iframes created by script.

basuke opened the pull request:
#### 8d3bcb122bc19a1aaa1c15fa1d6b2165d771a669
<pre>
[ResourceMonitor] Add support of monitoring for iframes created by script.
<a href="https://bugs.webkit.org/show_bug.cgi?id=284639">https://bugs.webkit.org/show_bug.cgi?id=284639</a>
<a href="https://rdar.apple.com/137691088">rdar://137691088</a>

Reviewed by Ben Nham.

If the iframe is created runtime, the script URL that creates the element is also the
target of checking the eligibility.

This patch add support for this case by checking the stack trace when iframe is created
and keep the URL of the known source in HTMLIFrameElement object. Then when document of
the frame is created, ResourceMonitor is also checks those URL.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitorIfExists):
(WebCore::Document::resourceMonitor):
(WebCore::Document::protectedResourceMonitor):
(WebCore::Document::parentResourceMonitorIfExists):
(WebCore::Document::setResourceMonitor): Deleted.
(WebCore::Document::parentResourceMonitor): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::HTMLIFrameElement):
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::commitData):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveResponse):
(WebCore::ResourceLoader::didReceiveBuffer):
(WebCore::ResourceLoader::resourceMonitorIfExists):
(WebCore::ResourceLoader::resourceMonitor): Deleted.
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::create):
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::setDocumentURL):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::ResourceMonitor::parentResourceMonitorIfExists const):
(WebCore::ResourceMonitor::parentResourceMonitor const): Deleted.
* Source/WebCore/loader/ResourceMonitor.h:

Canonical link: <a href="https://commits.webkit.org/287897@main">https://commits.webkit.org/287897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0625c9ed9f77cdfd50376455a1ec902f97adb7c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81047 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/572 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85576 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32033 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/590 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8384 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63266 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21030 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/350 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73766 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43564 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/247 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27927 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30491 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87011 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8277 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5852 "Found 2 new test failures: http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set.html storage/indexeddb/modern/workers-enable.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71569 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8454 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69601 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70805 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17672 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14848 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13774 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8238 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13761 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8075 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11595 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9883 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1409,8 +1409,13 @@ template<typename NameType>
-    if (LIKELY(element))
+    if (LIKELY(element)) {
+#if ENABLE(CONTENT_EXTENSIONS)
+        if (RefPtr iframe = dynamicDowncast<HTMLIFrameElement>(element.get()))
```
This is a hot code path, seems bad to do a branch for all element types when this is only for iframes (which are not constructed as frequently as other nodes). Why isn't this in the HTMLIFrameElement constructor? This is what `HTMLElementFactory::createKnownElement()` calls above for iframes anyway.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11135,20 +11140,27 @@ void Document::securityOriginDidChange()
+ResourceMonitor& Document::resourceMonitor()
+{
+    if (!m_resourceMonitor)
+        m_resourceMonitor = ResourceMonitor::create(*frame());
```
Should we asserting that this is not the top document?

basuke replied:
Added

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11135,20 +11140,27 @@ void Document::securityOriginDidChange()
+    return resourceMonitor();
 }
 
 ResourceMonitor* Document::parentResourceMonitor()
```
Should use `IfExists` since that's the pattern you're adopting elsewhere in your PR (though I personally prefer the `T* foo()` / `T& ensureFoo()` style personally). At the very least you should be consistent with yourself.

basuke replied:
But it is documented so I should follow the guideline. :)

cdumez commented on Source/WebCore/html/HTMLIFrameElement.h:
```diff
@@ -61,6 +61,11 @@ class HTMLIFrameElement final : public HTMLFrameElementBase {
 
+#if ENABLE(CONTENT_EXTENSIONS)
+    const URL& initiatorSourceURL() const { return m_initiatorSourceURL; }
+    void setInitiatorSourceURL(URL&& url) { m_initiatorSourceURL = url; }
```
missing `WTFMove()`

cdumez commented on Source/WebCore/html/HTMLIFrameElement.h:
```diff
@@ -83,6 +88,9 @@ class HTMLIFrameElement final : public HTMLFrameElementBase {
 #endif
     std::unique_ptr<LazyLoadFrameObserver> m_lazyLoadFrameObserver;
+#if ENABLE(CONTENT_EXTENSIONS)
+    URL m_initiatorSourceURL;
```
Why do we need to store this? Why can we check `document().url()` ?

basuke replied:
`document().url()` is the frame's URL, but we want to know who create `iframe`. Even if the frame's url is not eligible for checking, it is possible the script is eligible and to honor the decision, we need this.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -76,6 +71,18 @@ void ResourceMonitor::setEligibility(Eligibility eligibility)
+    didReceiveResponse(m_frameURL, ContentExtensions::ResourceType::Document);
+
+    if (RefPtr iframe = dynamicDowncast<HTMLIFrameElement>(protectedFrame()->ownerElement())) {
+        if (const auto& url = iframe->initiatorSourceURL(); !url.isEmpty())
```
`const auto&` -> `auto&` is what we normally use.

basuke committed:
[ResourceMonitor] Add support of monitoring for iframes created by script.
https://bugs.webkit.org/show_bug.cgi?id=284639
rdar://137691088

Reviewed by Ben Nham.

If the iframe is created runtime, the script URL that creates the element is also the
target of checking the eligibility.

This patch add support for this case by checking the stack trace when iframe is created
and keep the URL of the known source in HTMLIFrameElement object. Then when document of
the frame is created, ResourceMonitor is also checks those URL.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resourceMonitorIfExists):
(WebCore::Document::resourceMonitor):
(WebCore::Document::protectedResourceMonitor):
(WebCore::Document::parentResourceMonitorIfExists):
(WebCore::Document::setResourceMonitor): Deleted.
(WebCore::Document::parentResourceMonitor): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/html/HTMLIFrameElement.cpp:
(WebCore::HTMLIFrameElement::HTMLIFrameElement):
* Source/WebCore/html/HTMLIFrameElement.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::commitData):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::didReceiveResponse):
(WebCore::ResourceLoader::didReceiveBuffer):
(WebCore::ResourceLoader::resourceMonitorIfExists):
(WebCore::ResourceLoader::resourceMonitor): Deleted.
* Source/WebCore/loader/ResourceLoader.h:
* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::create):
(WebCore::ResourceMonitor::ResourceMonitor):
(WebCore::ResourceMonitor::setEligibility):
(WebCore::ResourceMonitor::setDocumentURL):
(WebCore::ResourceMonitor::addNetworkUsage):
(WebCore::ResourceMonitor::checkNetworkUsageExcessIfNecessary):
(WebCore::ResourceMonitor::parentResourceMonitorIfExists const):
(WebCore::ResourceMonitor::parentResourceMonitor const): Deleted.
* Source/WebCore/loader/ResourceMonitor.h:

Canonical link: https://commits.webkit.org/287897@main
